### PR TITLE
create permission utils

### LINKED ISSUE #979: Only ask Storage permission on Android 5 and 6
Android 13 introduces new limitations for reading files: https://developer.android.com/about/versions/13/behavior-changes-13#granular-media-permissions

Need to check if there's a way to opt out of this or if Google effectively killed the ability to import backups when targeting Android 13.

(This is the kind of stuff that makes me hate Android development, Google's half-baked "security improvements" (what security, apps can't access files themselves the user has to select the file) where they just ignore lots of valid use cases)

### PR DESCRIPTION
**Pull request by Lukmannudin**
Hi CatimaLoyaty!

I'm pretty new in open source project activities,

This for fix #979

### CODE REVIEW

**Review by Altonss — CHANGES_REQUESTED** [review]
Somehow there is a conflict with the master branch that needs to be resolved before merging :)

**Comment by Lukmannudin** [comment]
> Somehow there is a conflict with the master branch that needs to be resolved before merging :)

Hi @Altonss, Yes, I'm not figure out before. Solved

File: app/src/main/java/protect/card_locker/ImportExportActivity.java
Comment (Lukmannudin) [reply app/src/main/java/protect/card_locker/ImportExportActivity.java:442]: Hi @TheLastProject I just know we can put the max sdk, thank you for the insight!

In my reproduce, In android 5 and 6 we don't need to request permission, just put to android manifest. Could we not remove request permission completely except in android manifet?

File: app/src/main/java/protect/card_locker/ImportExportActivity.java
Comment (Lukmannudin) [reply app/src/main/java/protect/card_locker/ImportExportActivity.java]: Okay got it, I already testing as well and its turns out never needs request permission

File: app/src/main/java/protect/card_locker/ImportExportActivity.java
Comment (TheLastProject) [reply app/src/main/java/protect/card_locker/ImportExportActivity.java]: > In my reproduce, In android 5 and 6 we don't need to request permission, just put to android manifest. Could we remove request permission completely except in android manifest?

In Android 5 you don't need to request it no, but in Android 6 you do. So the code to request definitely is still necessary.

File: app/src/main/AndroidManifest.xml
Comment (TheLastProject) [inline app/src/main/AndroidManifest.xml:8]: No camera makes Catima pretty useless

File: app/src/main/AndroidManifest.xml
Comment (TheLastProject) [inline app/src/main/AndroidManifest.xml:8]: These permissions are necessary on Android 5 and 6 for loading images in both the `ScanActivity` from gallery button or in the Photos part of `LoyaltyCardEditActivity` (see https://github.com/CatimaLoyalty/Android/issues/979#issuecomment-1283040802)

**Review by TheLastProject — CHANGES_REQUESTED** [review]

File: app/src/main/AndroidManifest.xml
Comment (Lukmannudin) [reply app/src/main/AndroidManifest.xml]: I'm sorry I didn't notice, I think its because bad merge. Already reverted

File: app/src/main/AndroidManifest.xml
Comment (Lukmannudin) [reply app/src/main/AndroidManifest.xml]: @TheLastProject reverted

File: app/src/main/java/protect/card_locker/ImportExportActivity.java
Comment (Lukmannudin) [reply app/src/main/java/protect/card_locker/ImportExportActivity.java]: @TheLastProject reverted

**Comment by TheLastProject** [comment]
Haven't had the time to test this yet but I'm noticing at first glance:
1. The check in `ScanActivity` is still missing.
2. The no longer used `onRequestPermissionsResult` in ImportExportActivity still exists.
3. Users will have to accept the permission and then click the action again, this is suboptimal especially because Android offers the option to respond to the permission being granted with `onRequestPermissionResult`. You'll want to move the int out of the `PermissionUtils.java` and and make that configurable so the class in question knows why it requested the permission in the `onRequestPermissionResult`. It would probably be the best if on newer Android the code fakes permission acceptance (by calling `onRequestPermissionResult` itself) so the codebase doesn't have to care about the Android version, just knows "I will call the PermissionsUtil and then I'll get onRequestPermissionResult"
4. The maxSdk is still missing, so the permission is still marked as required on all Android versions (just never asked on newer ones)

**Comment by Lukmannudin** [comment]
> 

Hi @TheLastProject, thank you, your explanation is detail. Please take review again about this PR

**Review by TheLastProject — CHANGES_REQUESTED** [review]
Noticing some issues. Most importantly, these changes break importing completely on newer Android versions

File: app/src/main/java/protect/card_locker/ImportExportActivity.java
Comment (Lukmannudin) [reply app/src/main/java/protect/card_locker/ImportExportActivity.java]: done

File: app/src/main/AndroidManifest.xml
Comment (Lukmannudin) [reply app/src/main/AndroidManifest.xml]: done

File: app/src/main/java/protect/card_locker/utils/PermissionUtils.java
Comment (Lukmannudin) [reply app/src/main/java/protect/card_locker/utils/PermissionUtils.java]: @TheLastProject done, and already tested in my side on android 23 and 30 and also on real device on samsung s10+

File: app/src/main/java/protect/card_locker/ImportExportActivity.java
Comment (Lukmannudin) [reply app/src/main/java/protect/card_locker/ImportExportActivity.java]: @TheLastProject done

File: app/src/main/java/protect/card_locker/LoyaltyCardEditActivity.java
Comment (Lukmannudin) [reply app/src/main/java/protect/card_locker/LoyaltyCardEditActivity.java:1197]: Sorry I dont get it, so what your suggestion? @TheLastProject

Comment (Lukmannudin) [reply app/src/main/java/protect/card_locker/LoyaltyCardEditActivity.java:1197]: @TheLastProject  but I think is different, because Permission storage is for pick photos from files but permission PERMISSION_REQUEST_CARD_IMAGE_FROM_FILE_FRONT is from camera right?

Comment (TheLastProject) [reply app/src/main/java/protect/card_locker/LoyaltyCardEditActivity.java:1197]: It has `FROM_FILE` in the name, so no. And it's not about the name, it's about using a different variable depending on which view it is for. But looking more closely I see now that `mRequestedImage` is the same kind of ugly variable that `itemClickedView` is and I should probably clean that up first.

I don't think I'll still have time to clean this up before Hacktoberfest ends, so I'll just mark your change hacktoberfest-accepted as it's a helpful contribution even though I can't merge it just yet. I'll use it later to fix this :)

Comment (Lukmannudin) [reply app/src/main/java/protect/card_locker/LoyaltyCardEditActivity.java:1197]: hi @TheLastProject thank you so much to label as hacktoberfest-accepted, Okay got it I will wait to your commit first related this

**Comment by TheLastProject** [comment]
Your work was quite useful, I've refactored stuff a bit and turned it into #1186. Thanks again for your help :)

**Pull request closed without merging**
